Импортируем библиотеки

In [1]:
import pandas as pd
import numpy as np

Импортируем датасеты

In [2]:
train_data = pd.read_csv('data/train.csv')
lectures_data = pd.read_csv('data/lectures.csv')
questions_data = pd.read_csv('data/questions.csv')

Рассмотрим данные обучающей выборки

In [17]:
train_data.head()

row_id  timestamp  user_id  content_id  content_type_id  task_container_id  \
0       0          0      115        5692                0                  1   
1       1      56943      115        5716                0                  2   
2       2     118363      115         128                0                  0   
3       3     131167      115        7860                0                  3   
4       4     137965      115        7922                0                  4   

   user_answer  answered_correctly  prior_question_elapsed_time  \
0            3                   1                          NaN   
1            2                   1                      37000.0   
2            0                   1                      55000.0   
3            0                   1                      19000.0   
4            1                   1                      11000.0   

  prior_question_had_explanation  
0                            NaN  
1                          False  
2                          False  
3                          False  
4                          False

В первой же строке видим, что данные предыдущего вопроса отсутствуют (т.к. его не могло быть) - данные упорядочены во времени

Выведем информацию о наборе переменных в train

In [19]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101230332 entries, 0 to 101230331
Data columns (total 10 columns):
 #   Column                          Dtype  
---  ------                          -----  
 0   row_id                          int64  
 1   timestamp                       int64  
 2   user_id                         int64  
 3   content_id                      int64  
 4   content_type_id                 int64  
 5   task_container_id               int64  
 6   user_answer                     int64  
 7   answered_correctly              int64  
 8   prior_question_elapsed_time     float64
 9   prior_question_had_explanation  object 
dtypes: float64(1), int64(8), object(1)
memory usage: 7.5+ GB


Видим, что формат данных может быть сокращен для экономии памяти

Проведем необходимые изменения (условия были даны в задании)

In [20]:
train_data['user_id'] = train_data['user_id'].astype('int32')
train_data['content_id'] = train_data['content_id'].astype('int16')
train_data['content_type_id'] = train_data['content_type_id'].astype('int8')
train_data['task_container_id'] = train_data['task_container_id'].astype('int16')
train_data['user_answer'] = train_data['user_answer'].astype('int8')
train_data['answered_correctly'] = train_data['answered_correctly'].astype('int8')
train_data['prior_question_elapsed_time'] = train_data['prior_question_elapsed_time'].astype('float32')
train_data['prior_question_had_explanation'] = train_data['prior_question_had_explanation'].astype('float32')

Проверяем, как были отформатированы данные

In [21]:
train_data.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101230332 entries, 0 to 101230331
Data columns (total 10 columns):
 #   Column                          Non-Null Count      Dtype  
---  ------                          --------------      -----  
 0   row_id                          101230332 non-null  int64  
 1   timestamp                       101230332 non-null  int64  
 2   user_id                         101230332 non-null  int32  
 3   content_id                      101230332 non-null  int16  
 4   content_type_id                 101230332 non-null  int8   
 5   task_container_id               101230332 non-null  int16  
 6   user_answer                     101230332 non-null  int8   
 7   answered_correctly              101230332 non-null  int8   
 8   prior_question_elapsed_time     98878794 non-null   float32
 9   prior_question_had_explanation  100837826 non-null  float32
dtypes: float32(2), int16(2), int32(1), int64(2), int8(3)
memory usage: 3.3 GB


Объем использованной памяти сократился больше, чем вдвое (7.5+ -> 3.3 GB)

На самом деле в этом датасете есть 2 булевых/dummy/фиктвных переменных: content_type_id и prior_question_had_explanation

Из таблицы видно, что в двух последних переменных есть пропущенные значения. Выведем их точное количество-

In [22]:
train_data.isna().sum()

row_id                                  0
timestamp                               0
user_id                                 0
content_id                              0
content_type_id                         0
task_container_id                       0
user_answer                             0
answered_correctly                      0
prior_question_elapsed_time       2351538
prior_question_had_explanation     392506
dtype: int64

In [23]:
train_data.describe()

row_id     timestamp       user_id    content_id  \
count  1.012303e+08  1.012303e+08  1.012303e+08  1.012303e+08   
mean   5.061517e+07  7.703644e+09  1.076732e+09  5.219605e+03   
std    2.922268e+07  1.159266e+10  6.197163e+08  3.866359e+03   
min    0.000000e+00  0.000000e+00  1.150000e+02  0.000000e+00   
25%    2.530758e+07  5.243436e+08  5.408116e+08  2.063000e+03   
50%    5.061517e+07  2.674234e+09  1.071781e+09  5.026000e+03   
75%    7.592275e+07  9.924551e+09  1.615742e+09  7.425000e+03   
max    1.012303e+08  8.742577e+10  2.147483e+09  3.273600e+04   

       content_type_id  task_container_id   user_answer  answered_correctly  \
count     1.012303e+08       1.012303e+08  1.012303e+08        1.012303e+08   
mean      1.935222e-02       9.040624e+02  1.376123e+00        6.251644e-01   
std       1.377596e-01       1.358302e+03  1.192896e+00        5.225307e-01   
min       0.000000e+00       0.000000e+00 -1.000000e+00       -1.000000e+00   
25%       0.000000e+00       1.040000e+02  0.000000e+00        0.000000e+00   
50%       0.000000e+00       3.820000e+02  1.000000e+00        1.000000e+00   
75%       0.000000e+00       1.094000e+03  3.000000e+00        1.000000e+00   
max       1.000000e+00       9.999000e+03  3.000000e+00        1.000000e+00   

       prior_question_elapsed_time  prior_question_had_explanation  
count                 9.887879e+07                    1.008378e+08  
mean                  2.542384e+04                    8.894043e-01  
std                   2.122839e+04                    4.078948e-01  
min                   0.000000e+00                    0.000000e+00  
25%                   1.600000e+04                    1.000000e+00  
50%                   2.100000e+04                    1.000000e+00  
75%                   2.966600e+04                    1.000000e+00  
max                   3.000000e+05                    1.000000e+00

Рассмотрим сначала только события-вопросы (content_type_id == 0)

Создадим маску с соответствующим фильтром

Выведем на экран справку по датафрейму, убрав лишние параметры

In [24]:
train_data_q_only = (train_data['content_type_id'] == 0)
train_data[train_data_q_only].describe().loc[:, ['timestamp', 'answered_correctly' , 'prior_question_elapsed_time', 'prior_question_had_explanation']]

timestamp  answered_correctly  prior_question_elapsed_time  \
count  9.927130e+07        9.927130e+07                 9.887879e+07   
mean   7.702058e+09        6.572355e-01                 2.542385e+04   
std    1.159498e+10        4.746335e-01                 2.122839e+04   
min    0.000000e+00        0.000000e+00                 0.000000e+00   
25%    5.195280e+08        0.000000e+00                 1.600000e+04   
50%    2.669611e+09        1.000000e+00                 2.100000e+04   
75%    9.929147e+09        1.000000e+00                 2.966600e+04   
max    8.742577e+10        1.000000e+00                 3.000000e+05   

       prior_question_had_explanation  
count                    9.887879e+07  
mean                     9.070255e-01  
std                      4.119157e-01  
min                      0.000000e+00  
25%                      1.000000e+00  
50%                      1.000000e+00  
75%                      1.000000e+00  
max                      1.000000e+00

Выведены 3 временных показателя и параметр, отражающий правильность ответа студента (answered_correctly). Согласно его среднему, 65,7% ответов были верными
90,7% вопросов имели объяснение по предыдущему вопросу перед постановкой

Создадим копию нашего датасета, в котором уберем все лишние параметры, кроме "правильных ответов" и отметки о том, что предыдущий вопрос был объяснен

Посмотрим зависимость правильных ответов от получения разъяснений на предыдущий вопрос

In [25]:
train_df_answ = train_data[train_data_q_only].drop(['row_id',
                                                    'timestamp',
                                                    'user_id',
                                                    'content_id',
                                                    'content_type_id',
                                                    'task_container_id',
                                                    'user_answer',
                                                   'prior_question_elapsed_time'], axis=1)
train_df_answ.value_counts(normalize=True)

answered_correctly  prior_question_had_explanation
1                   1.0                               0.610599
0                   1.0                               0.296426
1                   0.0                               0.046546
0                   0.0                               0.046429
Name: proportion, dtype: float64

Как видно из таблицы, в 90,73 % (61,08+29,65) вопросов имели объяснение по предыдущему вопросу перед постановкой, как и было сказано выше
При этом только в 61,06 случае из 90,73 (или в 67,32%) учащийся, получивший разъяснения, отвечал правильно 

К тому же ответу можно прийти, отфильтровавшись по переменной prior_question_had_explanation

In [26]:
train_df_answ[(train_data['prior_question_had_explanation'] == 1.0)]['answered_correctly'].mean()

C:\Users\avmar\AppData\Local\Temp\ipykernel_33772\3982896269.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  train_df_answ[(train_data['prior_question_had_explanation'] == 1.0)]['answered_correctly'].mean()


0.6731885043701572

Или

In [27]:
train_data[train_data_q_only & (train_data['prior_question_had_explanation'] == 1.0)]['answered_correctly'].mean()

0.6731885043701572

Рассмотрим обратный случай: в какой доле случаев был назван правильный ответ, не смотря на то, что на предыдущий вопрос не давалось объяснения

In [28]:
train_data[train_data_q_only & (train_data['prior_question_had_explanation'] == 0)]['answered_correctly'].mean()

0.5006289408058143

Перезапишем датафрейм train_df_answ, где будет храниться копия исходного датафрейма с отбором по вопросам

In [29]:
train_df_answ = train_data[train_data_q_only].copy()

На основе параметра prior_question_had_explanation, создадим в нём ещё один столбец со сдвигом индекса на -1

Таким образом, в новом столбце будет информация о том, давались ли поянения к текущему вопросу

Назовем этот параметр соответствующим именем 'current_question_has_explanation'

In [30]:
train_df_answ['current_question_has_explanation'] = train_df_answ['prior_question_had_explanation'].shift(periods=-1, fill_value=0)

Посмотрим как теперь выглядит датафрейм

In [31]:
train_df_answ.head()

row_id  timestamp  user_id  content_id  content_type_id  task_container_id  \
0       0          0      115        5692                0                  1   
1       1      56943      115        5716                0                  2   
2       2     118363      115         128                0                  0   
3       3     131167      115        7860                0                  3   
4       4     137965      115        7922                0                  4   

   user_answer  answered_correctly  prior_question_elapsed_time  \
0            3                   1                          NaN   
1            2                   1                      37000.0   
2            0                   1                      55000.0   
3            0                   1                      19000.0   
4            1                   1                      11000.0   

   prior_question_had_explanation  current_question_has_explanation  
0                             NaN                               0.0  
1                             0.0                               0.0  
2                             0.0                               0.0  
3                             0.0                               0.0  
4                             0.0                               0.0

In [32]:
train_df_answ.tail()

row_id  timestamp     user_id  content_id  content_type_id  \
101230327  101230327  428564420  2147482888        3586                0   
101230328  101230328  428585000  2147482888        6341                0   
101230329  101230329  428613475  2147482888        4212                0   
101230330  101230330  428649406  2147482888        6343                0   
101230331  101230331  428692118  2147482888        7995                0   

           task_container_id  user_answer  answered_correctly  \
101230327                 22            0                   1   
101230328                 23            3                   1   
101230329                 24            3                   1   
101230330                 25            1                   0   
101230331                 26            3                   1   

           prior_question_elapsed_time  prior_question_had_explanation  \
101230327                      18000.0                             1.0   
101230328                      14000.0                             1.0   
101230329                      14000.0                             1.0   
101230330                      22000.0                             1.0   
101230331                      29000.0                             1.0   

           current_question_has_explanation  
101230327                               1.0  
101230328                               1.0  
101230329                               1.0  
101230330                               1.0  
101230331                               0.0

Убедимся, что мы работали только с вопросами: проверим, какие значения есть в переменой content_type_id

In [33]:
train_df_answ['content_type_id'].value_counts()

content_type_id
0    99271300
Name: count, dtype: int64

Проверим, как часто студент смотрел объяснение задачи в зависимости от правильности ответа

In [34]:
train_df_answ[(train_df_answ['answered_correctly'] == 1)]['current_question_has_explanation'].mean()

0.9291617

In [35]:
train_df_answ[(train_df_answ['answered_correctly'] == 0)]['current_question_has_explanation'].mean()

0.8644599

Видим, что после правильного ответа учащиеся чаще (92,92%) смотрели разъяснения, чем после неправильного (86,45%)

In [41]:
train_df_answ[(train_df_answ['answered_correctly'] == 1)]['prior_question_elapsed_time'].mean()/1000

25.30994921875

In [ ]:
25.31 секунда в среднем требовалась для правильного ответа

In [42]:
train_df_answ[(train_df_answ['answered_correctly'] == 0)]['prior_question_elapsed_time'].mean()/1000

25.64198828125

25.64 секунда в среднем требовалась для неверного ответа

Посмотрим на другие наборы данных

In [35]:
lectures_data.head()

lecture_id  tag  part           type_of
0          89  159     5           concept
1         100   70     1           concept
2         185   45     6           concept
3         192   79     5  solving question
4         317  156     5  solving question

questions_data.head()

Объединим датасет train с questions

In [4]:
train_df = train_data.merge(right=questions_data, how='left', left_on=['content_id'], right_on=['question_id'])
display(train_df)

row_id  timestamp     user_id  content_id  content_type_id  \
0                  0          0         115        5692                0   
1                  1      56943         115        5716                0   
2                  2     118363         115         128                0   
3                  3     131167         115        7860                0   
4                  4     137965         115        7922                0   
...              ...        ...         ...         ...              ...   
101230327  101230327  428564420  2147482888        3586                0   
101230328  101230328  428585000  2147482888        6341                0   
101230329  101230329  428613475  2147482888        4212                0   
101230330  101230330  428649406  2147482888        6343                0   
101230331  101230331  428692118  2147482888        7995                0   

           task_container_id  user_answer  answered_correctly  \
0                          1            3                   1   
1                          2            2                   1   
2                          0            0                   1   
3                          3            0                   1   
4                          4            1                   1   
...                      ...          ...                 ...   
101230327                 22            0                   1   
101230328                 23            3                   1   
101230329                 24            3                   1   
101230330                 25            1                   0   
101230331                 26            3                   1   

           prior_question_elapsed_time prior_question_had_explanation  \
0                                  NaN                            NaN   
1                              37000.0                          False   
2                              55000.0                          False   
3                              19000.0                          False   
4                              11000.0                          False   
...                                ...                            ...   
101230327                      18000.0                           True   
101230328                      14000.0                           True   
101230329                      14000.0                           True   
101230330                      22000.0                           True   
101230331                      29000.0                           True   

           question_id  bundle_id  correct_answer  part        tags  
0               5692.0     5692.0             3.0   5.0         151  
1               5716.0     5716.0             2.0   5.0         168  
2                128.0      128.0             0.0   1.0  131 149 92  
3               7860.0     7860.0             0.0   1.0  131 104 81  
4               7922.0     7922.0             1.0   1.0  131 149 92  
...                ...        ...             ...   ...         ...  
101230327       3586.0     3586.0             0.0   5.0           8  
101230328       6341.0     6341.0             3.0   5.0          60  
101230329       4212.0     4212.0             3.0   5.0          95  
101230330       6343.0     6343.0             0.0   5.0          60  
101230331       7995.0     7995.0             3.0   5.0          80  

[101230332 rows x 15 columns]

Разобъем столбец с тэгами тем на строки

In [ ]:
train_df['tags'] = train_df['tags'].str.split(',')
train_df = train_df.explode('tags')

Выведем наиболее популярные теги в зависимости от корректности ответа на вопрос

In [ ]:
train_df[(train_df['answered_correctly'] == 1)]['tags'].mode()
train_df[(train_df['answered_correctly'] == 0)]['tags'].mode()

Видим, что на вопросы с тэгом=24 получено всего 38.6% правильных ответов, а на вопросы с тэгом=68 более 86%

Аналогично рассматриваем данные для частей (part)

In [ ]:
train_df[(train_df['answered_correctly'] == 1)]['part'].mode()
train_df[(train_df['answered_correctly'] == 0)]['part'].mode()

Видим, что на вопросы с темой=5 получено 61% правильных ответов, а на вопросы с темой=1 более 74%

Мой код не работает из-за нехватки памяти. Я не знаю, как можно это исправить, не прибегая к сторонним библиотекам, которые мы не проходили